src/services/stringCompletions.ts

```diff
@@ -297,6 +310,22 @@ function convertStringLiteralCompletions(
     }
 }
 
+function getStringLiteralCompletionFromTemplateLiteralTypeAndTextOfNode(type: TemplateLiteralType, tokenTextContent: string, quoteChar: CharacterCodes.backtick | CharacterCodes.singleQuote | CharacterCodes.doubleQuote) {

```

Comment:
There is an existing algorithm for matching strings against template literal types that you can find in `isTypeMatchedByTemplateLiteralType`. Instead of trying to implement a custom one, I'd recommend looking into how you could reuse that one. You can expose *some* checker functionality on the `TypeChecker` using new methods annotated with `/** @internal */`.

Perhaps you will have to focus on the algorithm in `inferFromLiteralPartsToTemplateLiteral`. It could return a number for the last segment matched successfully instead of `undefined`. Using this number you could look at the text after it and suggest that as autocompletion. 

Note that this is just a rough idea, I don't know how feasible it will be in practice 😉 

It could be nice to use the same algorithm to autocomplete properties coming from index signatures with template literal keys. You don't have to implement this as part of this PR though. One thing at a time 😅

Reply:
Hmm, this matching works in the escaped string (source code) space, while the checker's works in raw string content space. Shouldn't be the biggest issue though, instead of escaping the type.text, we would need to unescape token.content. Especially for future generalisations, this is probably a good direction to take.

Reply:
Also, `inferFromLiteralPartsToTemplateLiteral` starts by first checking that both the starts and the ends of the source and the target are compatible, which is not what we need here.

Reply:
Actually, I don't think that function is very useful for matching of the partial string. I thing that this requires a different algorithm by nature.

Reply:
Could u expand on that? At the very least the rules there could guide u how to handle number interpolations and more

Reply:
Sure, there are functions like `isValidTypeForTemplateLiteralPlaceholder` that could certainly prove usefule, but I'm not sure the overall algorithm that finds the target text parts in the source type is that reuseable. And

> it's much less clear to me how they should work

meant less "I don't know how to do it" and more "I don't know what it should be doing". Like in

 ```ts
const x: `a${number}` = "a/**/
```
what is the completion? I guess nothing? How about
 ```ts
const x: `a${number}` = "ab/**/
```
Maybe replacement of `"ab` with `"a` ?